docs/string_grader.md

```diff
@@ -1,66 +1,236 @@
+
+When a response doesn't satisfy the given pattern, there are three types of feedback that you can provide, controlled by the `explain_validation` flag:
+
+* The student receives an error message. This does not consume an attempt (`explain_validation='err'`, default)
+* The student is graded incorrect, but receives a message. This does consume an attempt (`explain_validation='msg'`)
+* The student is graded incorrect, and no explanation is given. This does consume an attempt (`explain_validation=None`)
+
+In the first two cases, the message provided is given by the `invalid_msg` flag, which defaults to `Your input is not in the expected format`.
+
+Here is an example of using a validation pattern to accept inputs that look like chemical formulae for organic molecules. Note that anything that matches the pattern will be accepted.
+
+```pycon
+>>> grader = StringGrader(
+...     validation_pattern=r'([CNOH](_[0-9])?)+',
+...     explain_validation='msg',
+...     invalid_msg='Write a chemical formula containing hydrogen, oxygen, carbon and/or nitrogen',
+...     strip_all=True,   # Removes all spaces from the input
+...     accept_any=True
+... )
+>>> grader(None, 'NH_3') == {'grade_decimal': 1,
+...                          'msg': '',
+...                          'ok': True}
+True
+>>> grader(None, 'KCl') == {'grade_decimal': 0,
+...                          'msg': 'Write a chemical formula containing hydrogen, oxygen, carbon and/or nitrogen',
+...                          'ok': False}
+True
 
-```python
-grader = StringGrader(
-    answers={'expect': '', 'grade_decimal': 1, 'msg': 'Your answer has been recorded.'},
-    accept_any=True
-)
 ```
 
-Note that when either of these options are set to True, it is meaningless to provide more than one answer to check against, as they will all be graded as correct. Also note that you do not need to provide any answer at all in this situation.
+Here, we use validation to ensure that the student input matches the desired format before comparing to the answer, and give the student an error message if their input doesn't match the specification.
+
+```pycon
+>>> grader = StringGrader(
+...     answers='(1)(2)',
+...     validation_pattern=r'\([0-9]\)\([0-9]\)',
+...     explain_validation='err',
+...     strip_all=True   # Removes all spaces from the input
+... )
+>>> grader(None, '(1)(2)') == {'grade_decimal': 1, 'msg': '', 'ok': True}
+True
+>>> grader(None, '(2)(1)') == {'grade_decimal': 0, 'msg': '', 'ok': False}
+True
+>>> grader(None, '(a)(2)')
+Traceback (most recent call last):
```

Comment:
Sadly this doesn't work in Python 3. You can change it if you want, or you can consider that my problem 😛 

Python 3 prints traceback ends with `mitxgraders.exceptions.InvalidInput` instead of `InvalidInput`.

See http://python3porting.com/problems.html#handling-expected-exceptions

(Possibly this could be resolved by modifing sys. excepthook, but this seems hard and I can't find anyone else who's done it online.)

Reply:
Thanks :)

Reply:
Should be fixed now, I think...